felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -2550,37 +2568,58 @@ func (d *bpfEndpointManagerDataplane) applyPolicyToWeps(
 	return ingressAP, egressAP, errors.Join(ingressErr, egressErr)
 }
 
-func (d *bpfEndpointManagerDataplane) applyPolicyToDataIface(
+func (d *bpfEndpointManagerDataplane) attachTCProgramToDataIface(
 	ifaceName string,
 	ep *proto.HostEndpoint,
 	state *bpfInterfaceState,
-	ap *tc.AttachPoint,
-	apxdp *xdp.AttachPoint,
-) (*tc.AttachPoint, *tc.AttachPoint, *xdp.AttachPoint, error) {
+	ap *tc.AttachPoint) (*tc.AttachPoint, *tc.AttachPoint, error) {
 	ingressAttachPoint := *ap
 	egressAttachPoint := *ap
-	xdpAttachPoint := *apxdp
 
+	var ingressErr, egressErr error
 	var parallelWG sync.WaitGroup
 	var ingressAP, egressAP *tc.AttachPoint
-	var xdpAP *xdp.AttachPoint
-	var ingressErr, egressErr, xdpErr error
-
-	parallelWG.Add(2)
+	parallelWG.Add(1)
 	go func() {
 		defer parallelWG.Done()
 		ingressAP, ingressErr = d.attachDataIfaceProgram(ifaceName, ep, PolDirnIngress, state, &ingressAttachPoint)
 	}()
 
+	egressAP, egressErr = d.attachDataIfaceProgram(ifaceName, ep, PolDirnEgress, state, &egressAttachPoint)
+	parallelWG.Wait()
+	return ingressAP, egressAP, errors.Join(ingressErr, egressErr)
+}
+
+func (d *bpfEndpointManagerDataplane) applyPolicyToDataIface(
+	ifaceName string,
+	ep *proto.HostEndpoint,
+	state *bpfInterfaceState,
+	ap *tc.AttachPoint,
+	apxdp *xdp.AttachPoint,
+	ifaceType IfaceType,
+	attachXDPOnly bool,
+) (*tc.AttachPoint, *tc.AttachPoint, *xdp.AttachPoint, error) {
+	xdpAttachPoint := *apxdp
+
+	var parallelWG sync.WaitGroup
+	var xdpAP *xdp.AttachPoint
+	var ingressAP, egressAP *tc.AttachPoint
+	var xdpErr, tcErr error
+
+	parallelWG.Add(1)
```

Comment:
woudn't it be easier to handle it in the original function, where you would decide whether to add the xdp goroutine in the waitgroup if you start it or not? Now you may have a routine that only starts to finish the wait (confusing) and does nothing and then you have a second waitgroup ...

Reply:
I feel handling doing this here is easier to handle.